### PR TITLE
Catch synchronous errors in `Doc._otApply`

### PR DESCRIPTION
**Pull request by alecgibson**
Fixes https://github.com/share/sharedb/issues/257

As outlined in the above issue, calling `Doc.submitOp` can sometimes
`throw` a synchronous error, for example when submitting an invalid op
that causes `type.apply` to `throw`.

This is surprising behaviour, because an error handler should already be
provided in the callback supplied to `submitOp`. What's more, the doc
could be left partially mutated and out-of-sync with the server, which
could lead to confusing behaviour.

This change wraps all internal usage of `Doc._otApply` in `try`/`catch`
blocks. If an error is thrown in this method, then we attempt to
recover with `Doc._hardRollback`, which will attempt to reset the
document and then call the callbacks of the pending ops with the error.

This change also prevents `Doc._hardRollback` from emitting an error if
at least one pending op callback was invoked, so that we don't branch
error handling behaviour.

### CODE REVIEW

File: lib/client/doc.js [hunk 1]
```diff
@@ -889,7 +900,7 @@ Doc.prototype._hardRollback = function(err) {
   this.fetch(function() {
     var called = op && callEach(op.callbacks, err);
     for (var i = 0; i < pending.length; i++) {
-      callEach(pending[i].callbacks, err);
+      called = callEach(pending[i].callbacks, err) || called;
```
Comment (alecgibson) [inline lib/client/doc.js]: @nateps @ericyhwang do you know if this change will have any weird side-effects? Originally it would not count a pending callback as having called something, so it would both call back, _and_ emit, which seems a bit odd?

It's forced some updates to a couple of edge case tests, which I'll highlight also. I'm assuming they're okay to change...? In them, we actually assert that the callback is called _and_ an error is emitted; is that the expected behaviour?

File: test/client/submit.js [hunk 2]
```diff
@@ -628,19 +628,12 @@ describe('client submit', function() {
         if (err) return done(err);
         doc2.del(function(err) {
           if (err) return done(err);
-          doc.pause();
-          var calledBack = false;
-          doc.on('error', function() {
```
Comment (alecgibson) [inline test/client/submit.js]: Is this emission actually expected, given the callback is called? (See above comment)

File: test/client/submit.js [hunk 3]
```diff
@@ -657,19 +650,12 @@ describe('client submit', function() {
           if (err) return done(err);
           doc2.create({age: 5}, function(err) {
             if (err) return done(err);
-            doc.pause();
-            var calledBack = false;
-            doc.on('error', function() {
```
Comment (alecgibson) [inline test/client/submit.js]: Is this emission actually expected, given the callback is called? (See above comment)

**Comment by alecgibson** [comment]
Also apologies for the code coverage decrease, but I don't have any idea how to reach the other `catch` blocks without either:

  - directly calling the internal methods (which feels like cheating)
  - abusing middleware to mutate `doc.data` after `_submitOp`, but before `_handleOp` (which also feels like cheating)

**Comment by alecgibson** [comment]
Okay - code coverage back up I think. I did actually manage to work out some crazy contrived (but theoretically possible) way that `type.apply` could throw after the op was successfully submitted.

**Comment by alecgibson** [comment]
As discussed, we should check that **every** op has a callback. If any of them don't, then we should emit an error (so that no errors are ever swallowed).

**Comment by alecgibson** [comment]
@ericyhwang @nateps I've updated the hard rollback callback logic. I've actually tweaked it a tiny bit more than discussed - the inflight op is no longer treated as "special" relative to other pending ops. Here's the commit message for clarity:

We previously followed the following logic for callbacks in a hard
rollback:

  - check there's an inflight op, and that it has a callback
  - if this isn't the case, then emit the error

This logic doesn't work very well in the situation where we don't have
an inflight op. This can happen in the following case:

  - I submit an invalid op
  - the op is added to the pending ops queue
  - we attempt to apply the op to the local document before flushing
  - `type.apply` throws, so the op is never flushed from pending to
    inflight
  - we perform a hard rollback
  - we now find we have no inflight op, but we do have a pending op

In this case, since we have no inflight op, we'd emit the error, but
also call the pending op callback, causing both a callback call _and_
an error emission, which is a bit surprising.

This change updates our logic, so that:

  - we check that there are some ops to callback
  - and that all of those ops have callbacks

If there are no callbacks, or if any of the ops don't handle the error,
then we emit, so that we can be sure of never swallowing an error. We
also avoid treating the inflight op any differently to pending ops, so
that we avoid behaviour that's dependent on where in its lifecycle an op
throws.

However, if the client handles all of their op submission errors, then
we never emit.
